benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,254 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-24, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+// RAJA/RAJA.hpp cannot be included here because the include logic will
+// default all atomic implementations to a desul backend.
+#include "RAJA/policy/loop/policy.hpp"
+#include "RAJA/policy/openmp/atomic.hpp"
+#include "RAJA/RAJA.hpp"
+#include "RAJA/policy/openmp/policy.hpp"
+
+#include <type_traits>
+#include <iostream>
+#include <sstream>
+
+/// This helper template is used to deduce if device memory allocations are necessary
+/// inside the body of the benchmark, using the type of the execution policy.
+template<typename>
+struct IsGPU : public std::false_type {};
```

Comment:
You should be able to get a resource based on the policy and use it to allocate memory and synchronize.

Reply:
I see what you mean about getting the resource based on the policy but I don't understand how this would simplify the allocation logic

Reply:
Resources have allocation and deallocation functions on them.

Reply:
I think what @MrBurmark is referring to is that you don't need the tag structs, etc. to determine how you manage memory and synchronization. You do these things generically using the resource, like we do in our tests, where there is no back-end specific code. For example, https://github.com/LLNL/RAJA/blob/develop/test/functional/forall/segment/tests/test-forall-RangeSegment.hpp#L20